src/matlab2tikz.m

```diff
@@ -963,23 +884,107 @@ case guitypes()
+if isInvisibleHG1orOctave || isInvisibleHG2;
     legendhandle = [];
-    switch getEnvironment
-        case 'Octave'
-            % Make sure that m2t.legendHandles is a row vector.
-            for lhandle = m2t.legendHandles(:)'
-                ud = get(lhandle, 'UserData');
-                if isVisibleContainer(lhandle) && any(handle == ud.handle)
-                    legendhandle = lhandle;
-                    break;
-                end
+end
+end
+% ==============================================================================
+function entries = getLegendEntries(m2t)
+entries = [];
+legendHandle = m2t.axesContainers{end}.LegendHandle;
+
+if isempty(legendHandle), return, end
+
+switch getEnvironment
+    case 'Octave'
+        % See set(hlegend, "deletefcn", {@deletelegend2, ca, [], [], t1, hplots}); in legend.m
+        delfun  = get(legendHandle,'deletefcn');
+        entries = delfun{6};
+        
+        % Bubble-up legend entry properties from child to hggroup root
+        % for guessable objects
+        guessable = {'specgraph.errorbarseries','matlab.graphics.chart.primitive.ErrorBar',...
+                     'specgraph.scattergroup'  ,'matlab.graphics.chart.primitive.Scatter'};
+        for ii = 1:numel(entries)
+            child = entries(ii);
+            anc   = ancestor(child,'hggroup');
+            if isempty(anc)
+                continue
             end
-        case 'MATLAB'
-            % no action needed
-        otherwise
-            errorUnknownEnvironment();
-    end
+            cl = guessOctavePlotType(anc);
+            if any(strcmpi(cl, guessable))
+                legendString = get(child,'displayname');
+                set(anc,'displayname',legendString);
+                entries(ii) = anc;
+            end
+        end
+        
+    case 'MATLAB'
+        % Undocumented property (exists at least since 2008a)
+        entries = double(get(legendHandle,'PlotChildren'));
+        
+        % Take only the first child from a pure hggroup (e.g. bodeplots)
+        for ii = 1:numel(entries)
+            entry     = entries(ii);
+            isHggroup = any(strcmpi(get(entry,'Type'),{'hggroup', 'matlab.graphics.primitive.Group'})) & ...
+                            strcmpi(class(handle(entry)),'hggroup');
+            if isHggroup
+                children    = get(entry, 'Children');
+                firstChild  = children(1);
+                % Inherits DisplayName from hggroup root
+                set(firstChild, 'DisplayName', get(entry, 'DisplayName'));
+                entries(ii) = firstChild;
+            end
+        end
+end
+end
+% ==============================================================================
+function string = getLegendString(m2t, h)
+entries = m2t.axesContainers{end}.LegendEntries;
+idx     = ismember(entries, h);
+string  = getOrDefault(entries(idx), 'displayname', '');
+
+% HG1: autogenerated legend strings, i.e. data1,..., dataN, do not populate
```

Comment:
This happens when you do something like `plot([1:10],[1:10]); legend('show')` (or even spawns a lot of lines when the `y` parameter is a matrix).